### PR TITLE
Smooth derivative directions, add function options

### PR DESCRIPTION
**Pull request by rchristie**

### CODE REVIEW

**Comment by rchristie** [comment]
Users will need update in ABI-Software/mapclientplugins.meshgeneratorstep#28.

File: src/scaffoldmaker/utils/derivativemoothing.py [hunk 1]
```diff
@@ -56,6 +56,22 @@ def getLastArcLength(self):
...
+    def getDelta(self):
+        '''
+        Caller must have called evaluateArcLength!
+        '''
+        x1 = self._parameters[0]
+        x2 = self._parameters[2]
+        componentsCount = len(x1)
+        return [ (x2[c] - x1[c]) for c in range(componentsCount) ]
+
+    def getExpresson(self, expressionIndex):
```
Comment (mahyar-osn) [inline src/scaffoldmaker/utils/derivativemoothing.py:68]: Typo: `getExpression`

File: src/scaffoldmaker/utils/derivativemoothing.py [hunk 2]
```diff
@@ -257,19 +273,19 @@ def addEdgeCurve(self, element, eft, basisLocalNodeFunctions, basisLocalNodes, d
...
-                derivativeEdge = (edge, expressionIndex, math.fabs(scaleFactor))
+                derivativeEdge = (edge, expressionIndex, totalScaleFactor)
                 derivativeEdges = self._derivativeMap.get(derivativeKey)
                 if derivativeEdges:
                     derivativeEdges.append(derivativeEdge)
                 else:
                     self._derivativeMap[derivativeKey] = [ derivativeEdge ]
 
-    def smooth(self, maxIterations=10, arcLengthTolerance=1.0E-6):
+    def smooth(self, updateDirections=False,maxIterations=10, arcLengthTolerance=1.0E-6):
```
Comment (mahyar-osn) [inline src/scaffoldmaker/utils/derivativemoothing.py:288]: Add space between `updateDirections=False, maxIterations=10`.

File: src/scaffoldmaker/utils/derivativemoothing.py [hunk 3]
```diff
@@ -323,16 +348,39 @@ def smooth(self, maxIterations=10, arcLengthTolerance=1.0E-6):
...
-                        if (mag <= 0.0):
-                            print('Node', nodeIdentifier, 'label', nodeValueLabel, 'version', nodeVersion, 'has negative mag', mag)
-                        x = setMagnitude(x, mag)
+                        otherExpressionIndex = 3 if (expressionIndex == 1) else 1
+                        otherd = edge.getParameter(otherExpressionIndex)
+                        if updateDirections:
+                            thisx = edge.getParameter(expressionIndex - 1)
+                            otherx = edge.getParameter(otherExpressionIndex - 1)
+                            bothEndsOnBoundary = False
+                            otherExpression = edge.getExpresson(otherExpressionIndex)
```
Comment (mahyar-osn) [inline src/scaffoldmaker/utils/derivativemoothing.py:360]: `getExpresson` Should automatically be fixed as above.

**Review by mahyar-osn — CHANGES_REQUESTED** [review]
Just minor changes. Otherwise, all good. Reviewed and tested.

**Review by mahyar-osn — APPROVED** [review]
Looks and works great!

**Pull request merged**